### ISSUE TITLE
Cannot build kotlin bindings on Windows

### ISSUE THREAD

**Issue by kusius** [n0-computer/iroh-ffi#84]
Branch; `feat-kotlin`
Platform: `windows`
NDK: `26.1.10909125`

Hi all! Figured I'd try to build the kotlin bindings today.

-  I have added the `aarch64-linux-android` target with `rustup`
 - NDK is present and is in path 
 - Noticed that in `.cargo/config.toml` the archiver used is not present in the prebuilt NDK toolchain but there was a `llvm-ar` shipped within that toolchain, so I modified the line to `ar = "llvm-ar"`
 
 I'm getting a weird failure when the build reaches the `redb` crate compilation:

```rust
> cargo build --target aarch64-linux-android --release
   Compiling redb v1.1.0
error: linker `aarch64-linux-android29-clang` not found
  |
  = note: program not found

error: could not compile `redb` (lib) due to previous error
```
The curious thing is that `aarch64-linux-android29-clang`  is on the path and works fine ( I'm guessing it worked for all the other packages that where compiled before `redb`) 

Any direction would be appreciated !

**Comment by dignifiedquire** [comment]
hey, thanks for trying it out, I so far have only been testing this on a mac, and there the instructions I added in the PR were enough. 

One thing to test would be trying to build an example from uniffi-rs (eg https://github.com/mozilla/uniffi-rs/tree/main/examples/arithmetic) to see if it depends on the iroh code, or the general setup.

**Comment by kusius** [comment]
Allrighty, success (both building the uniffi arithmetic example and iroh-ffi) but once again Windows and tools made for it, show its hideous face. I apologise for the lost comment about to follow, but I figure the info can be useful !

**TL;DR  The NDK toolchain shipped by google for windows does not play well with rust.**

## Issue 1: `aarch64-linux-android29-clang` not found on windows

About `aarch64-linux-android29-clang` not being on the path, found out through reading [this](https://users.rust-lang.org/t/cross-compiling-for-android-from-windows-using-clang-fails-with-could-not-exec-the-linker-1-is-not-a-valid-win32-application/16076/3), that the referenced `aarch64-linux-android29-clang` is not an executable, it is supposed to be a shell script. On the windows NDK toolchain this is a separate `aarch64-linux-android29-clang.cmd` located in the path of the toolchain along with the clang compiler/linker `aarch64-linux-android29-clang` 

> Changing `aarch64-linux-android29-clang` to `aarch64-linux-android29-clang.cmd`  in  `.cargo/config.toml`  fixed the initial issue 😃 

## Issue 2: `aarch64-linux-android-ar` not found on windows

Again an NDK toolchain error, there is no such thing as a `aarch64-linux-android-ar` but there is an `llvm-ar` shipped with the NDK. 

> Renaming that executable in the toolchain fixed this issue 

## Issue 3: But wait ... there's (still) Windows

With both tools found compilation was ok, but there was a specific point where the `aarch64-linux-android29-clang.cmd` batch script invoked could not parse the version script path argument given to it. Fortunately I've found a fix mentioned in [this ](https://github.com/rust-lang/rust/issues/113711) rust-lang issue. This allowed the compiler arguments to be parsed correctly and the target `.so` to be built

## My conclusion and maybe some suggestions 

Both fixes were relevant in some way or another to the NDK toolchain shipped by google for windows. I believe there's nothing wrong with neither `iroh-ffi` or `uniffi-rs`.

There is however the option I would suggest of trying out the `cargo-ndk` rust extension to build the target `cdylib`s for android if it works flawlessly for Mac and Linux platforms as well.

Also the kotlin section in the Readme.md maybe should more explicitly state that the [Kotlin Command Line Tools](https://kotlinlang.org/docs/tutorials/command-line.html) should be installed on the machine (I was lucky because it was already installed and present in my path) , this is also stated as a requirement in `uniffi-rs`.

Happy to contribute it any more ways I can bit by bit, bringing iroh to android !

**Comment by dignifiedquire** [comment]
Thanks a lot for this, and very glad to hear you got it to workd. 

> There is however the option I would suggest of trying out the `cargo-ndk` rust extension to build the target `cdylib`s for android if it works flawlessly for Mac and Linux platforms as well.

Definitely going to take a look at this, thanks for the suggestion.


> Happy to contribute it any more ways I can bit by bit, bringing iroh to android !

Yes please, both help on expanding on the documentation and any feedback if the bindings actually work they way they are supposed to would be much appreciated.

**Comment by kusius** [comment]
Yes, next step will be to try them out
